design/design.md

```diff
@@ -0,0 +1,223 @@
+to the `CompletionQueue` (each `CallData` object has a pointer to the same `CompletionQueue`).
+* `doProcess()`, which is called when a request is received, passes to the JobQueue
+a coroutine that calls the appropriate handler. `doProcess()` also creates
+another `CallData` object to listen for this request type, since this object
+is processing a response and no longer listening.
+* `doFinish()`, which is called when the response has been sent,
+destroys `this` object
+
+In the setup of the event loop, a `CallData` object for each rpc method is
+created. When an event occurs, `CompletionQueue::Next()` returns a pointer to a
+`CallData` object, and the event loop calls `Proceed()` on the returned object. 
+If the event was reception of a request, `Proceed()` will call `doProcess()`. If
+the event was a response was sent, `Proceed()` will call `doFinish()`.
+
+```
+  void EventLoop() {
+    // Spawn new CallData instances to serve new clients.
+    new AccountInfoCallData(&service_, completion_queue_.get(), app_);
+    new FeeCallData(&service_, completion_queue_.get(), app_);
+    ///... Additional CallData instances for additional rpc methods
+
+    void* tag;  // uniquely identifies a request.
+    bool ok;
+    while (true) {
+      // Block waiting to read the next event from the completion queue. The
+      // event is uniquely identified by its tag, which is the memory address
+      // of a CallData instance.
+      completion_queue_->Next(&tag, &ok);
+      static_cast<CallData*>(tag)->Proceed();
+    }
+  }
+```
+
+There will be a handler for each grpc method, separate from the existing json
+handlers. The handler will take in the protobuf object that represents the
+request, and return a protobuf object that represents the response. The protobuf
+object will be a member of a templated class `RPC::ContextGeneric<T>`, where T
+is the type of the protobuf object. `RPC::ContextGeneric<T>` is very similar to
+`RPC::Context`, except the `params` member is of type `T` and there are no
+`Headers`.
+
+Below is an example handler signature; `io::xpring` is the namespace of the
+protobuf objects, whereas `GetFeeRequest` is the request type and `Fee` is the
+response type.
+```
+io::xpring::Fee doFee(RPC::ContextGeneric<io::xpring::GetFeeRequest>& context);
+```
+Versioning for grpc is usually done by changing the package name, which creates
+protobuf objects in a new c++ namespace. For example, changing the package name
+to `io.xpring.v2` results in objects with namespace `io::xpring::v2`. In the event
+that a handler does not change between versions (which implies the objects are
+the same in all but namespace), the old handler can be reused by templating the
+handler signature like so:
+
+```
+template<class T, class R>
+R doFee(RPC::ContextGeneric<T>& context);
+```
+## Use Case
+The GRPC service will be used in a request response format (as opposed to streaming).
```

Comment:
delete this. use case is not needed if the reader doesn't know what gRPC is used for we are wasting time here.